File: Array/Easy/GCD of Strings/C++ Solution/GCD_of_Strings.cpp
#include <iostream>
#include <string>

using namespace std;


class Solution {
    public:
        string gcdOfStrings(string str1, string str2) {
            if (str1.length() < str2.length()) {
                return gcdOfStrings(str2, str1);
            }

            //base case
            if (str1 == str2) {
                return str1;
            }

            if (startsWith(str1, str2)) {
                return gcdOfStrings(str1.substr(str2.length()), str2);
            }

            return "";
        }

        bool startsWith(string str1, string str2) {
            return (str1.substr(0, str2.length()) == str2);
        }
};


int main() {
    Solution solution_class;
    string str1, str2;
    cout << "Enter str1: ";
    cin >> str1;

    cout << "Enter str2: ";
    cin >> str2;

    string result = solution_class.gcdOfStrings(str1, str2);
    cout << result;

    return 0;
}

File: Array/Easy/GCD of Strings/Python Solution/gcdOfStrings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        if len(str2) > len(str1):
            return self.gcdOfStrings(str2, str1)
        
        if str1 == str2:
            return str1
        
        if str1.startswith(str2):
            return self.gcdOfStrings(str1[len(str2):], str2)
        
        return ""

str1 = input("Enter str1: ")
str2 = input("Enter str2: ")
solution_class = Solution()
print(solution_class.gcdOfStrings(str1, str2))